DB/src/main/java/io/deephaven/db/util/config/MutableInputTable.java

```diff
@@ -36,7 +46,50 @@
     TableDefinition getTableDefinition();
 
     /**
-     * Write newData to this table.
+     * Helper to check if a table is compatible with this table, so that it could be added as contents.
+     *
+     * @param tableToApply the table to check if it can used to add or modify this input table
+     * @throws TableDefinition.IncompatibleTableDefinitionException if the definitions are not compatible
+     */
+    default void validateAddOrModify(final Table tableToApply) {
+        getTableDefinition().checkMutualCompatibility(tableToApply.getDefinition());
+    }
+
+    /**
+     * Validates that the given table definition is suitable to be passed to {@link #delete(Table)}.
+     *
+     * @param tableToDelete the definition of the table to delete
+     * @throws UnsupportedOperationException if this table does not support deletes
+     * @throws io.deephaven.db.exceptions.ArgumentException if the given definition isn't compatible to be used to
+     *         delete
+     */
+    default void validateDelete(Table tableToDelete) {
+        final TableDefinition keyDefinition = tableToDelete.getDefinition();
+        final TableDefinition thisDefinition = getTableDefinition();
+        final StringBuilder error = new StringBuilder();
+        final List<String> keyNames = getKeyNames();
+        for (String keyColumn : keyNames) {
+            final ColumnDefinition<?> colDef = keyDefinition.getColumn(keyColumn);
+            final ColumnDefinition<?> thisColDef = thisDefinition.getColumn(keyColumn);
+            if (colDef == null) {
+                error.append("Key Column \"").append(keyColumn).append("\" does not exist.\n");
+            } else if (!colDef.isCompatible(thisColDef)) {
+                error.append("Key Column \"").append(keyColumn).append("\" is not compatible.\n");
+            }
+        }
+        final List<String> extraKeys = keyDefinition.getColumnNames().stream().filter(kd -> !keyNames.contains(kd))
+                .collect(Collectors.toList());
+        if (!extraKeys.isEmpty()) {
+            error.append("Unknown key columns: ").append(extraKeys);
```

Comment:
Do lists append correctly?

Reply:
To my knowledge, all Lists have a toString that calls toString on children (unlike java arrays), so yes, this will list the contents? Is that not the case?

This is the same code as before, just moved out to the interface so that consumers like the grpcapi can directly ask if an operation is legal rather than relying on "guess and find out"